### DISCUSSION TITLE
Issue with Line profiles and crystal shapes

### DISCUSSION THREAD

**Discussion by Jesusmunozmontecinos** [mtex-toolbox/mtex#660] [Ask Anything]
Dear all,

First at all, sorry for these silly questions... my knowledge at computing is really basic

I'm having problems at plotting misorientation angle histograms considering a line profile along several JADEITE grains (not a single grain)
After follow the example from https://mtex-toolbox.github.io/EBSDProfile.html as:
```matlab
[grains,ebsd('indexed').grainId,ebsd.mis2mean] = calcGrains(ebsd('indexed'));
ebsd(grains(grains.grainSize<=40)) = [];
[grains,ebsd('indexed').grainId,ebsd.mis2mean] = calcGrains(ebsd('indexed'),'angle',10*degree);
F = halfQuadraticFilter;
ebsdS = smooth(ebsd('indexed'),F,'fill',grains);
[grains,ebsdS('indexed').grainId,ebsdS.mis2mean] = calcGrains(ebsdS('indexed'));
lineSec =  [3656   928; 1 928];
ebsd_line = spatialProfile(ebsd('Jadeite'),lineSec)
plot(ebsd_line('Jadeite').y,angle(ebsd_line.orientations,ebsd_line('Jadeite').orientations)/degree) 
%I know that here is the problem, however I have tried many combinations adding 'Jadeite' Id with no success.
```
What is the correct way for doing such plot?

I take advantage of this topic since I'm likely falling in the same issue when plotting crystal shapes (as in this example : https://mtex-toolbox.github.io/EBSDTutorial.html)
```matlab
cS = crystalShape.clinopyroxene(ebsd('Jadeite').CS) %The problem is already here, I'm lost on how using this function

% select only grains with more then 100 pixels
grains = grains(grains.grainSize > 100);

% plot at the positions of the Forsterite grains the crystal shapes
hold on
plot(grains('Jadeite'),0.7*cS,'FaceColor',[0.3 0.5 0.3])
hold off
```
Wishing you a happy new year and a good 2021,

Jesús

**Comment by ralfHielscher** [comment]
Hi Jesus,

in the line
```matlab
plot(ebsd_line('Jadeite').y,angle(ebsd_line.orientations,ebsd_line('Jadeite').orientations)/degree) 
```
you probably forgot `('Jadeite')` in the second occurrence. Appart from this it makes not so much sense to compute angle of an orientation with itself - it will be always zero.

In the docu-page you referred to we have
```matlab
angle(ebsd_line(1:end-1).orientations,ebsd_line(2:end).orientations)/degree)
```
which computed the angle between one orientation on the line to the previous orientation on the line.

Ralf.

**Comment by Jesusmunozmontecinos** [comment]
Thanks for your answer Ralf, but ?'m still facing troubles

I tried with the following:
```matlab
plot(ebsd_line('Jadeite').y,...
angle(ebsd_line('Jadeite',1).orientations,ebsd_line('Jadeite').orientations)/degree) 
% this part of the code is still not working, I guess there is a problem with the introduction of 'Jadeite' Id somewhere that I can not realize.

% misorientation gradient
hold all
plot(0.5*(ebsd_line.y('Jadeite',1:end-1)+ebsd_line.y('Jadeite',2:end)),...
angle(ebsd_line('Jadeite',1:end-1).orientations,ebsd_line('Jadeite',2:end).orientations)/degree) % this part seems to work, however, when I tried to compute this alone, no information is appearing on the diagram

hold off

xlabel('y'); ylabel('misorientation angle in degree')

legend('to reference orientation','orientation gradient')
```
cordially,

**Reply by ralfHielscher** [reply]
What about
```matlab
plot(ebsd_line.y, angle(ebsd_line(1).orientations,ebsd_line.orientations)/degree) 
```
Ralf.

**Comment by Jesusmunozmontecinos** [comment]
Ralf,
I'm still in troubles,
I tried with:
plot(ebsd_line.y, angle(ebsd_line(1).orientations,ebsd_line.orientations)/degree)  

and also adding 'Jadeite' everywhere is possible and it did not work,

sorry for disturbing with this

**Reply by ralfHielscher** [reply]
Posting error messages as well as the output of `ebsd_line` would be extremely helpful

**Comment by Jesusmunozmontecinos** [comment]
Dear Ralf,

the output of ebsd_line is :
>> ebsd_line
 
ebsd_line = EBSD (show methods, plot)
 
 Phase  Orientations  Mineral         Color  Symmetry  Crystal reference frame
     2     476 (59%)  Jadeite  DarkSeaGreen     12/m1        X||a*, Y||b, Z||c
     3     332 (41%)  Jadeite     Goldenrod     12/m1        X||a*, Y||b, Z||c
 
 Properties: bands, bc, bs, error, mad, x, y, grainId, mis2mean
 Scan unit : um

The error message of computing plot(ebsd_line.y, angle(ebsd_line(1).orientations,ebsd_line.orientations)/degree) is (even if writing 'Jadeite' everywhere I think is possible):

>> plot(ebsd_line.y, angle(ebsd_line(1).orientations,ebsd_line.orientations)/degree)
Error using phaseList/checkSinglePhase (line 301)

----------------------------------------------------------------
Your variable contains the phases: Jadeite, Jadeite

However, your are executing a command that is only permitted for a single phase!

Please see modify EBSD data for how to restrict EBSD data or grains to a single phase.
----------------------------------------------------------------


Error in phaseList/get.CS (line 154)
      id = checkSinglePhase(pL);

Error in EBSD/get.orientations (line 167)
        ori = orientation(ebsd.rotations,ebsd.CS);

Error in EBSD/subsref (line 40)
  [varargout{1:nargout}] = builtin('subsref',ebsd,s);

**Reply by kilir** [reply]
Hi Jesus,
you have two times Jadeite, in your data? 
- If those are different, you need to make sure that you'd give them different names, e.g. in your definition if CS when importing your data.
- If they should both be the same but you created them by accident (e.g. by merging different datasets), you can try to change the phase e.g. `ebsd(ebsd.phase==3).phase=2`.

Cheers,
Rüdiger

**Comment by Jesusmunozmontecinos** [comment]
Dear Ralf and Rüdiger,

after applying ebsd(ebsd.phase==3).phase=2 , the error message disappear, however, my results are not satisfying :

Code-->
ebsd(ebsd.phase==3).phase=2
[grains,ebsd('indexed').grainId,ebsd.mis2mean] = calcGrains(ebsd('indexed'));
ebsd(grains(grains.grainSize<=40)) = [];
[grains,ebsd('indexed').grainId,ebsd.mis2mean] = calcGrains(ebsd('indexed'),'angle',10*degree);
F = halfQuadraticFilter;
ebsdS = smooth(ebsd('indexed'),F,'fill',grains);
[grains,ebsdS('indexed').grainId,ebsdS.mis2mean] = calcGrains(ebsdS('indexed'));
lineSec =  [2000   928; 1000 928];
ebsd_line = spatialProfile(ebsdS('Jadeite'),lineSec)

plot(ebsdS('Jadeite'),ebsdS('Jadeite').mis2mean.angle./degree)
mtexColorbar
hold on
plot(grains.boundary,'edgecolor','k','linewidth',.5)
hold off
CLim(gcm,[0 10])
mtexColorMap hot
line(lineSec(:,1),lineSec(:,2),'linewidth',2)
hold on
figure (2)
plot(angle(ebsd_line(1:end-1,'Jadeite').orientations,ebsd_line(2:end,'Jadeite').orientations)/degree)
hold off

I got this:

![image](https://user-images.githubusercontent.com/76631957/104233130-9f804580-5451-11eb-9d3a-e304cd35284a.png)

With respect to the line profile:
- The output (right hand figure) is somehow strange for me. In the misorientation plot we see little misorientation along the profile except the yellowish crystals at around the middle of the segment, where I expect a jump on the misorientation profile with a wide of some pixels. However, the only thin I see are some peaks that do not correspond to what I see in the map.
- The X axis of the profile do not correspond to the length established in the lineSec =  [2000   928; 1000 928];, which I assume is 1000 (pixels, points?).
- Why do I have such big jumps at >100 degrees in misorientation while in the misorientation map I see 40° max (see below for the output of this figure without the CLim(gcm,[0 10]))

![image](https://user-images.githubusercontent.com/76631957/104233806-c9863780-5452-11eb-9940-a874f77210b2.png)


Sorry for disturbing so much with this,

Hopping you a good week,

Jesús

**Comment by kilir** [comment]
Hi jesus,
1) the angles you colorcode in your plot (angle of misorientation to grain mean orientation) are mostly unrelated to what you plot on your line plot (angle of misorientation between neigboring pixels, also across grain boundaries)
2) on figure 2 you do not specify what to plot on your x-axis, hence Matlab just takes numbers of measurements. Why not use distance from origin (additional output from `spatialProfile`)
3) the large angles relate to misorientation angles at your grain boundaries; also see (1)

Hope that helps.

Cheers,
Rüdiger

**Comment by Jesusmunozmontecinos** [comment]
Dear Rüdiger,

Thanks a lot for your answer,
Finally I got what I was looking for!,

Cheers,

Jesús
